### ISSUE TITLE
Issues with advanced enum and classproperty cases

### ISSUE THREAD

**Issue by ngnpope** [astral-sh/ty#3680]
### Summary

I have a pattern that I use where I have class properties on an enum to be able to provide subsets of members that fit in a certain category and then have properties on the member that can be used to determine whether they are within those subsets. This is quite useful for keeping locality of behaviour without having to maintain separate global sets of enum members on which to perform containment checks.

Currently this typically works rather well for me when using mypy, but there are a few issues that crop up with ty that would be good to fix.

To keep things simple, I've written up an example below with details of what the different type checkers do and some thoughts throughout. Specifically for ty, it seems as though there are the following issues:

- The classproperty return type on the enum results in ``expected `tuple[Self@canines, ...]`, found `tuple[Literal[Animal.WOLF]]` `` (This is the thing causing me the most errors.)
- While looking into this I found that, although `Animal.canines` returns `(cls.WOLF,)`, and I expect this to be assignable to `tuple[Animal]` and `tuple[Animal, ...]`, ty doesn't complain about `tuple[Animal, Animal, Animal]`. All other type checkers get this right.

```python
from __future__ import annotations

import enum
from typing import TYPE_CHECKING, Literal, assert_type

if TYPE_CHECKING:
    from collections.abc import Callable
    from typing import Self


class classproperty[R]:  # noqa: N801
    """
    A minimal `@classproperty` descriptor decorator implementation.

    Similar to `django.utils.functional.classproperty()`.
    """

    def __init__(self, fget: Callable[..., R], /) -> None:
        """Initialize the classproperty for the provided method."""
        self.fget = fget

    def __get__(self, instance: object, owner: type[object], /) -> R:
        """Retrieves the value from the method, passing in the class."""
        return self.fget(owner)


def test_classproperty_basic() -> None:
    class Container:
        _value = "<class>"

        def __init__(self) -> None:
            self._value = "<instance>"

        @classproperty
        def source(cls) -> str:
            return cls._value

    assert Container.source == "<class>"
    assert Container().source == "<class>"


def test_classproperty_enum() -> None:
    class Animal(enum.StrEnum):
        LION = enum.auto()
        LYNX = enum.auto()
        WOLF = enum.auto()

        @classproperty
        def canines(cls) -> tuple[Self, ...]:
            # I don't think this should be an error, expecially as enums are final. It's notable
            # that mypy doesn't complain here, yet treats `Animal.canines` as `tuple[Animal]` and
            # ignores the explicit return annotation which should result in `tuple[Animal, ...]` as
            # every other type checker does. It feels like zuban is most friendly here.
            #
            # true negative: mypy, zuban
            # false positive: pyrefly, pyright, ty
            return (cls.WOLF,)  # pyrefly: ignore[bad-return] # pyright: ignore[reportReturnType] # ty: ignore[invalid-return-type]

        @classproperty
        def felines(cls) -> tuple[Self, ...]:
            # As above.
            #
            # true negative: mypy, zuban
            # false positive: pyrefly, pyright, ty
            return (cls.LION, cls.LYNX)  # pyrefly: ignore[bad-return] # pyright: ignore[reportReturnType] # ty: ignore[invalid-return-type]

        @enum.property
        def is_canine(self) -> bool:
            return self in self.canines

        @enum.property
        def is_feline(self) -> bool:
            return self in self.felines

    # Class properties should be accessible from the base class and members.
    assert Animal.canines == (Animal.WOLF,)
    assert Animal.felines == (Animal.LION, Animal.LYNX)
    assert Animal.LION.canines == (Animal.WOLF,)
    assert Animal.LION.felines == (Animal.LION, Animal.LYNX)

    # Other properties should be accessible from the members only, not the base class.
    assert not hasattr(Animal, "is_canine")
    assert not hasattr(Animal, "is_feline")
    assert Animal.LION.is_canine is False
    assert Animal.LION.is_feline is True
    assert Animal.WOLF.is_canine is True
    assert Animal.WOLF.is_feline is False

    # Standard, makes sense.
    #
    # true negative: mypy, pyrefly, pyright, ty, zuban
    assert_type(Animal.LION, Literal[Animal.LION])
    assert_type(Animal.LYNX, Literal[Animal.LYNX])
    assert_type(Animal.WOLF, Literal[Animal.WOLF])

    # Ok. This is just because mypy gets this wrong as it really should have `tuple[Animal, ...]`
    # and because ty treats `tuple[Animal]` as assignable to `tuple[Animal, ...]`. So I guess the
    # "false" positives are not really wrong and it's optional to support this.
    #
    # true negative: mypy, ty
    # false positive: pyrefly, pyright, zuban
    assert_type(Animal.canines, tuple[Animal])  # pyrefly: ignore[assert-type] # pyright: ignore[reportAssertTypeFailure] # zuban: ignore[misc]
    assert_type(Animal.felines, tuple[Animal, Animal])  # pyrefly: ignore[assert-type] # pyright: ignore[reportAssertTypeFailure] # zuban: ignore[misc]

    # Pretty convincing that mypy is wrong here - it's ignoring the explicit return type of the
    # property and using the type of the value returned and not complaining that it's incompatible.
    #
    # true negative: pyrefly, pyright, ty, zuban
    # false positive: mypy
    assert_type(Animal.canines, tuple[Animal, ...])  # type: ignore[assert-type]
    assert_type(Animal.felines, tuple[Animal, ...])  # type: ignore[assert-type]

    # Standard, makes sense.
    #
    # true negative: mypy, pyrefly, pyright, ty, zuban
    assert_type(Animal.LION.is_canine, Literal[False])
    assert_type(Animal.LION.is_feline, Literal[True])
    assert_type(Animal.WOLF.is_canine, Literal[True])
    assert_type(Animal.WOLF.is_feline, Literal[False])

    # Again, this is likely just because ty treats `Literal[False]` and `Literal[True]` as
    # assignable to `bool`. So I guess the "false" positives are not really wrong and it's optional
    # to support this.
    #
    # true negative: ty
    # false positive: mypy, pyrefly, pyright, zuban
    assert_type(Animal.LION.is_canine, bool)  # type: ignore[assert-type] # pyrefly: ignore[assert-type] # pyright: ignore[reportAssertTypeFailure] # zuban: ignore[misc]
    assert_type(Animal.LION.is_feline, bool)  # type: ignore[assert-type] # pyrefly: ignore[assert-type] # pyright: ignore[reportAssertTypeFailure] # zuban: ignore[misc]
    assert_type(Animal.WOLF.is_canine, bool)  # type: ignore[assert-type] # pyrefly: ignore[assert-type] # pyright: ignore[reportAssertTypeFailure] # zuban: ignore[misc]
    assert_type(Animal.WOLF.is_feline, bool)  # type: ignore[assert-type] # pyrefly: ignore[assert-type] # pyright: ignore[reportAssertTypeFailure] # zuban: ignore[misc]

    # This one is clearly something ty is getting wrong. `Animal.canines` returns `(cls.WOLF,)` so
    # `tuple[Animal]` can be assigned to `tuple[Animal, ...]` but it certainly shouldn't be
    # assignable to `tuple[Animal, Animal, Animal]`.
    #
    # true positive: mypy, pyrefly, pyright, zuban
    # false negative: ty
    assert_type(Animal.canines, tuple[Animal, Animal, Animal])  # type: ignore[assert-type] # pyrefly: ignore[assert-type] # pyright: ignore[reportAssertTypeFailure] # zuban: ignore[misc]
    assert_type(Animal.felines, tuple[Animal, Animal, Animal])  # type: ignore[assert-type] # pyrefly: ignore[assert-type] # pyright: ignore[reportAssertTypeFailure] # zuban: ignore[misc]
```

`pyproject.toml`

```toml
[tool.mypy]
strict = true

[tool.pyright]
enableTypeIgnoreComments = false
typeCheckingMode = "strict"

[tool.pyrefly]
enabled-ignores = ["pyrefly"]
preset = "strict"

[tool.ty.analysis]
respect-type-ignore-comments = false

[tool.zuban]
mode = "default"
strict = true
```

### Version

ty 0.0.44

**Comment by carljm** [comment]
Hi! Thanks for the report, lots to digest here.

The `invalid-return-type` are true positives, despite the finality of enums. The reason is because `Self` could be a specific enum literal type. This is something that we explicitly and intentionally support, and is useful for other scenarios. Given the `tuple[Self, ...]` return type as currently annotated, the correct inference for e.g. `Animal.LION.canines` would be `tuple[Literal[Animal.LION], ...]`, which is clearly wrong. I think `canines` and `felines` should instead be annotated to return `tuple[Animal, ...]` explicitly, eliminating that problem.

It looks like with the `-> tuple[Self, ...]` return annotation, we also don't resolve that through `@classproperty`, so we end up with `Unknown` return type for `canines` and `felines`. This is a bug, and already tracked in #2520. Switching to `-> tuple[Animal, ...]` works around it in this case.

We do currently model that `Animal.is_canine` exists, which apparently does not match the runtime behavior. We also don't get its return type correct, we just see it as `Any`. So it looks like we could improve our handling of `enum.property`. I filed https://github.com/astral-sh/ty/issues/3681 for at least getting the type right -- modeling "doesn't exist on the class" will be more complex, and seems relatively low priority.

Unfortunately, the rest of your test code is seeing misleading behaviors because of the `assert not hasattr(Animal, "is_canine")` line. Since we believe `Animal.is_canine` does exist, we treat that assertion as always-failing, and the rest of the code as unreachable (this is visible in an LSP or the playground, as we grey out the code). We generally don't emit diagnostics in unreachable code, so you aren't seeing our actual behavior on any of the following `assert_type` calls. It looks to me like once we account for the above bugs, and eliminate the reachability problem, all the remaining ty behaviors here are as I would expect.

Closing this since it's rather broad, and I think all the specific issues where we should improve are tracked in other issues now. Feel free to file an (ideally narrowly-focused, with minimal repro) issue for any remaining gaps that you see!

**Comment by ngnpope** [comment]
Hi. Sorry for the delayed response - was ill earlier this week.

> Unfortunately, the rest of your test code is seeing misleading behaviors...and the rest of the code as unreachable...

How embarrassing... Can't believe I missed that 😂 Many of the issues obviously just melted away and the checkers largely all then agree. What I should have posted is the following:

```python
from __future__ import annotations

import enum
from typing import TYPE_CHECKING, Literal, assert_type

if TYPE_CHECKING:
    from collections.abc import Callable
    from typing import Self


class classproperty[R]:  # noqa: N801
    """
    A minimal `@classproperty` descriptor decorator implementation.

    Similar to `django.utils.functional.classproperty()`.
    """

    def __init__(self, fget: Callable[..., R], /) -> None:
        """Initialize the classproperty for the provided method."""
        self.fget = fget

    def __get__(self, instance: object, owner: type[object], /) -> R:
        """Retrieves the value from the method, passing in the class."""
        return self.fget(owner)


def test_classproperty_basic() -> None:
    class Container:
        _value = "<class>"

        def __init__(self) -> None:
            self._value = "<instance>"

        @classproperty
        def source(cls) -> str:
            return cls._value

    assert Container.source == "<class>"
    assert Container().source == "<class>"


def test_classproperty_enum() -> None:
    class Animal(enum.StrEnum):
        LION = enum.auto()
        LYNX = enum.auto()
        WOLF = enum.auto()

        @classproperty
        def canines(cls) -> tuple[Self, ...]:
            return (cls.WOLF,)  # pyrefly: ignore[bad-return] # pyright: ignore[reportReturnType] # ty: ignore[invalid-return-type]

        @classproperty
        def felines(cls) -> tuple[Self, ...]:
            return (cls.LION, cls.LYNX)  # pyrefly: ignore[bad-return] # pyright: ignore[reportReturnType] # ty: ignore[invalid-return-type]

        @enum.property
        def is_canine(self) -> bool:
            return self in self.canines

        @enum.property
        def is_feline(self) -> bool:
            return self in self.felines

    assert_type(Animal.LION, Literal[Animal.LION])
    assert_type(Animal.LYNX, Literal[Animal.LYNX])
    assert_type(Animal.WOLF, Literal[Animal.WOLF])

    assert_type(Animal.canines, tuple[Animal, ...])  # ty: ignore[type-assertion-failure]
    assert_type(Animal.felines, tuple[Animal, ...])  # ty: ignore[type-assertion-failure]

    assert_type(Animal.LION.is_canine, bool)
    assert_type(Animal.LION.is_feline, bool)
    assert_type(Animal.WOLF.is_canine, bool)
    assert_type(Animal.WOLF.is_feline, bool)

    # Class properties should be accessible from the base class and members.
    assert Animal.canines == (Animal.WOLF,)
    assert Animal.felines == (Animal.LION, Animal.LYNX)
    assert Animal.LION.canines == (Animal.WOLF,)
    assert Animal.LION.felines == (Animal.LION, Animal.LYNX)

    # Other properties should be accessible from the members only, not the base class.
    assert not hasattr(Animal, "is_canine")
    assert not hasattr(Animal, "is_feline")
    assert Animal.LION.is_canine is False
    assert Animal.LION.is_feline is True
    assert Animal.WOLF.is_canine is True
    assert Animal.WOLF.is_feline is False
```

With that, it's as you mentioned:

> The invalid-return-type are true positives, despite the finality of enums...

I guess, in the case of enums, this makes sense that `Self` would refer to literals of members, but it's not immediately obvious. I didn't really see anything about it in the typing spec. Maybe something should be added to one or both of these sections?

- https://typing.python.org/en/latest/spec/enums.html
- https://typing.python.org/en/latest/spec/generics.html#self

> Switching to `-> tuple[Animal, ...]` works around it in this case.

With `-> tuple[Animal, ...]` all the remaining issues go away. It's just mypy/zuban that accept `-> tuple[Self, ...]`, so I guess they're incorrect?

**Comment by carljm** [comment]
I agree it would be good to better specify the behavior of `Self` with enum types. I think ty is the only type checker that does this; all other type checkers infer `Color` here, we infer `Literal[Color.RED]`:

```py
from typing import Self
from enum import Enum

class Color(Enum):
    RED = 1
    BLUE = 2
    
    def foo(self) -> Self:
        return self
    
reveal_type(Color.RED.foo())  # ty says `Literal[Color.RED]`, other type checkers say `Color`
```

I think this is a useful behavior and would not prefer to give it up. But I wouldn't say that mypy/zuban are wrong here; as long as they infer `Color` on the last line, it's sound for them to allow `return Color.BLUE` in `foo`. Since this is not specified, either behavior is conformant.

Oddly, pyright and pyrefly disallow `return Color.BLUE` in `foo`, but also don't solve `Self` to a precise literal type. That doesn't seem to me like a very useful combination of behaviors, but there may be reasons for it that I'm not aware of.
